sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java

```diff
@@ -198,32 +200,51 @@ public String getAuthorizationTokenUsingResourceTokens(Map<String, String> resou
     }
 
     private Mac getMacInstance() {
+        reInitializeIfPossible();
+
+        try {
+            return (Mac)this.macInstance.clone();
+        } catch (CloneNotSupportedException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    /*
+     * Ensures that this.macInstance is initialized
+     * In-case of credential change, optimistically will try to refresh the macInstance
+     *
+     * Implementation is non-blocking, the one which acquire the lock will try to refresh
+     * with new credentials
+     *
+     * NOTE: Calling it CTOR ensured that default is initialized.
+     */
+    private void reInitializeIfPossible() {
         int masterKeyLatestHashCode = this.credential.getKey().hashCode();
 
         //  Master key has changed, or this is the first time we are getting mac instance
         if (masterKeyLatestHashCode != this.masterKeyHashCode) {
-            synchronized (this.credential) {
-                if (masterKeyLatestHashCode != this.masterKeyHashCode) {
-                    byte[] masterKeyBytes = this.credential.getKey().getBytes(StandardCharsets.UTF_8);
-                    byte[] masterKeyDecodedBytes = Utils.Base64Decoder.decode(masterKeyBytes);
-                    SecretKey signingKey = new SecretKeySpec(masterKeyDecodedBytes, "HMACSHA256");
-                    try {
-                        Mac macInstance = Mac.getInstance("HMACSHA256");
-                        macInstance.init(signingKey);
-                        //  Update the master key hash code
-                        this.masterKeyHashCode = masterKeyLatestHashCode;
-                        this.macInstance = macInstance;
-                    } catch (NoSuchAlgorithmException | InvalidKeyException e) {
-                        throw new IllegalStateException(e);
+            // Try to acquire the lock, the one who got lock will try to refresh the macInstance
+            boolean lockAcquired = this.macInstanceLock.tryLock();
```

Comment:
what is the benefit of using ReentrantLock here? 

key rotation doesn't happen often and this adds a bit to complexity. 
Why aren't we simply using synchronized?

Reply:
reEntrant lock comes into picture only when there is change detected. Otherwise no impact, no?

Reply:
that's correct.

my question is: isn't `synchronized(.)` easier to use? what's the benefit of reEntrant lock here? I don't think you will get any benefit and its usage is slightly more complicated than using `synchronized(.)`

is there any benefit in using reEntrant lock over `synchrnozied()`

Reply:
synchronized is blocking, where as reEntry.TryLock is enabling non-blocking. So only one thread will go into re-initialization.

Reply:
if the key is rotated, then shouldn't the threads attempting to get a new macInstance wait and block till a new macInstance generated?

Is the old macInstance value returned to the threads which don't get the lock a valid macInstance in the key-rotation scenario?